engine/src/main/java/org/terasology/rendering/dag/RenderTaskListGenerator.java

```diff
@@ -102,41 +115,41 @@ private void logIntermediateRendererListForDebugging(List<Node> orderedNodes) {
 
         for (Node node : orderedNodes) {
             if (node.isEnabled()) {
-
                 if (logger.isInfoEnabled()) {
                     // Marker tasks just add a dividing line to the logger output
                     taskList.add(new MarkerTask(node.getClass().getSimpleName()));
                     enabledNodes++; // we count them only for statistical purposes
                     potentialTasks += 2 * node.getDesiredStateChanges().size() + 1;
                 }
 
-                // generating tasks for the desired state changes
+                // Examine the StateChanges requested by current Node, and insert/update them in persistentStateChanges.
                 for (StateChange currentStateChange : node.getDesiredStateChanges()) {
+                    StateChange persistentStateChange = persistentStateChanges.get(currentStateChange.getClass());
 
-                    // State changes persist beyond the node that request them if following nodes
-                    // require the exact same change.
-                    persistentStateChange = persistentStateChanges.get(currentStateChange.getClass());
-
-                    // for a state change to be necessary there can't be an identical one already persisting
-                    if (persistentStateChange == null || !currentStateChange.equals(persistentStateChange)) {
-                        // no persistent state change of this subType found: the state change is necessary (not redundant)
+                    // This exact StateChange was not requested (or was requested with a different value by the last Node),
+                    // so this StateChange is required and we insert it (or replace the old StateChange with it).
+                    if (!currentStateChange.equals(persistentStateChange)) {
                         stateChangesToAdd.add(currentStateChange);
                         persistentStateChanges.put(currentStateChange.getClass(), currentStateChange);
                     }
+                    // Else: The exact same StateChange is already persisting, so don't add it again.
 
-                    // set the state change as relevant for this iteration
+                    // Mark the StateChange as relevant for this iteration.
                     persistentStateChangesRelevant.put(currentStateChange.getClass(), true);
                 }
 
-                // set the state change as relevant for this iteration
-                for (Iterator<Map.Entry<Class<?>, StateChange>> iterator = persistentStateChanges.entrySet().iterator(); iterator.hasNext(); ) {
-                    Map.Entry<Class<?>, StateChange> stateChange = iterator.next();
+                // Iterate through all StateChanges
+                Iterator<Map.Entry<Class<?>, StateChange>> iterator = persistentStateChanges.entrySet().iterator();
```

Comment:
Same as last comment, since this is the only place this variable is used, can I leave this as it is?

I really don't want to go back to the C-days when all the variables had to be declared on top, overscoping them and making it hard to determine where they are used (provided they are).